language/move-prover/bytecode-to-boogie/src/prelude.bpl

```diff
@@ -0,0 +1,543 @@
+function {:constructor} Address(a: Address): Value;
+function {:constructor} ByteArray(b: ByteArray): Value;
+function {:constructor} Str(a: String): Value;
+function {:constructor} Vector(v: ValueArray): Value;
+const DefaultValue: Value;
+function {:builtin "MapConst"} MapConstValue(v: Value): [int]Value;
+
+type {:datatype} ValueArray;
+function {:constructor} ValueArray(v: [int]Value, l: int): ValueArray;
+const EmptyValueArray: ValueArray;
+axiom l#ValueArray(EmptyValueArray) == 0;
+axiom v#ValueArray(EmptyValueArray) == MapConstValue(DefaultValue);
+function {:inline} AddValueArray(a: ValueArray, v: Value): ValueArray {
+    ValueArray(v#ValueArray(a)[l#ValueArray(a) := v], l#ValueArray(a) + 1)
+}
+function {:inline} RemoveValueArray(a: ValueArray): ValueArray {
+    ValueArray(v#ValueArray(a)[l#ValueArray(a) := DefaultValue], l#ValueArray(a) - 1)
+}
+function {:inline} ConcatValueArray(a1: ValueArray, a2: ValueArray): ValueArray {
+    ValueArray(
+        (lambda i: int :: if i < l#ValueArray(a1) then v#ValueArray(a1)[i] else v#ValueArray(a2)[i - l#ValueArray(a1)]),
+        l#ValueArray(a1) + l#ValueArray(a2))
+}
+function {:inline} ReverseValueArray(a: ValueArray): ValueArray {
+    ValueArray(
+        (lambda i: int :: if 0 <= i && i < l#ValueArray(a) then v#ValueArray(a)[l#ValueArray(a) - i - 1] else DefaultValue),
+        l#ValueArray(a)
+    )
+}
+function {:inline} UpdateValueArray(a: ValueArray, i: int, elem: Value): ValueArray {
+    ValueArray(v#ValueArray(a)[i := elem], l#ValueArray(a))
+}
+function {:inline} SwapValueArray(a: ValueArray, i: int, j: int): ValueArray {
+    ValueArray(v#ValueArray(a)[i := v#ValueArray(a)[j]][j := v#ValueArray(a)[i]], l#ValueArray(a))
+}
+function {:inline} IsEmpty(a: ValueArray): bool {
+    l#ValueArray(a) == 0
+}
+
+function {:inline} HasType(tv: TypeValue, v: Value): bool {
+    if (is#BooleanType(tv))
+        then is#Boolean(v)
+    else if (is#IntegerType(tv))
+        then is#Integer(v)
+    else if (is#AddressType(tv))
+        then is#Address(v)
+    else if (is#ByteArrayType(tv))
+        then is#ByteArray(v)
+    else if (is#StrType(tv))
+        then is#Str(v)
+    else
+        // is#VectorType(tv) || is#StructType(tv) must be true
+        is#Vector(v)
+}
+function WellFormedValueArray(a: ValueArray): bool {
+    0 <= l#ValueArray(a) &&
+    (forall i: int :: (0 <= i && i < l#ValueArray(a)) || v#ValueArray(a)[i] == DefaultValue) &&
+    (forall i: int :: 0 <= i && i < l#ValueArray(a) ==> WellFormedValue(v#ValueArray(a)[i]))
+}
+function WellFormedValue(v: Value) : bool {
+    if (is#Vector(v))
+        then WellFormedValueArray(v#Vector(v))
+    else
+        true
+}
+function IsEqual(v1: Value, v2: Value): bool {
+    (v1 == v2) ||
+    (is#Vector(v1) &&
+     is#Vector(v2) &&
+     vlen(v1) == vlen(v2) &&
+     (forall i: int :: 0 <= i && i < vlen(v1) ==> IsEqual(vmap(v1)[i], vmap(v2)[i])))
+}
+
+function {:inline} vmap(v: Value): [int]Value {
+    v#ValueArray(v#Vector(v))
```

Comment:
How can I tell if the `v` here is bound to `v: Value` or the `v` in `function {:constructor} TypeValueArray(v: [int]TypeValue, l: int): TypeValueArray` or `function {:constructor} Vector(v: ValueArray): Value;`?

Reply:
# is not an operator.  v#ValueArray is a string. So is v#Vector.  Both of them are selector functions that are automatically generated by the Boogie parser for defined datatype constructors.